### PR TITLE
Hook up the terminal UI to the new Juju commands backend for the web cli.

### PR DESCRIPTION
**Pull request by hatched**
## Done

- Created a plugin for xterm for communicating with the Juju `/commands` endpoint for the web cli.
- Terminal accepts user input and displays the responses from Juju.

This feature is only available in a custom build of Juju and won't be enabled or available until Juju 2.9 is released. 
There are a number of edge case issues that I've been tracking and I'm not sure that this is the best UX for this feature but in any event this is just one step towards the ultimate goal.

## QA

- Build Juju from the `develop` branch.
- Bootstrap a controller.
- Get the user credentials by running `juju dashboard`.
  - Update line 46 in `Terminal.js` to those credential values.
- In the dashboard view a model and interact with the shell.

Or just ask me for a demo :)

## Details

Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1395

## Screenshots

Screen Shot 2020-09-08 at 4 52 30 PM

### CODE REVIEW

File: src/components/Terminal/Terminal.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import React, { useEffect, useRef, useState } from "react";
 import { useHistory } from "react-router-dom";
-import { useSelector } from "react-redux";
-import { getMacaroons } from "app/selectors";
+// import { useSelector } from "react-redux";
+// import { getMacaroons } from "app/selectors";
```
Comment (hatched) [inline src/components/Terminal/Terminal.js:4]: Typically I wouldn't comment out code being checked into the repo, but I think in this case it's important to leave this information here for the follow-ups on fetching the authentication credentials.

File: src/pages/Models/Details/ModelDetails.js [hunk 2]
```diff
@@ -95,6 +101,20 @@ const filterModelStatusData = (modelStatusData, appName) => {
   return modelStatusData;
 };
 
+const generateTerminalComponent = (modelUUID, controllerWSHost) => {
+  return null; // XXX Remove me to see the Terminal
```
Comment (hatched) [inline src/pages/Models/Details/ModelDetails.js:105]: The Terminal UI will be disabled until the 2.9 release of Juju and the dashboard.

File: src/components/Terminal/Terminal.js [hunk 3]
```diff
@@ -41,12 +41,16 @@ const Terminal = ({ address, modelName }) => {
         history.push(`/models/${model}`);
       }
     };
-    const creds = { macaroons };
+    // const creds = { macaroons };
+    // XXX credentials need to be pulled from the store.
+    const credentials = {
+      user: "admin",
+      credentials: "9eef4b03836b2a805dd5a03d43d5f3c0",
```
Comment (barrymcgee) [inline src/components/Terminal/Terminal.js:48]: This repo is public - is this a security risk?

File: src/components/Terminal/xterm-addon-juju.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    this._socket = socket;
+    this._buffer = "";
+    this._user = userData.user;
+    this._credentials = userData.credentials;
+    this._macaroons = userData.macaroons;
+  }
+
+  activate(terminal) {
+    this._disposables.push(
+      addSocketListener(this._socket, "message", (ev) => {
```
Comment (barrymcgee) [inline src/components/Terminal/xterm-addon-juju.js:13]: It's convention in this codebase that pass events using `(e)`

File: src/components/Terminal/Terminal.js (hunk 3 above)
Comment (hatched) [reply src/components/Terminal/Terminal.js:48]: Well now that you called it out it is!!!

Haha jk :) It's just the randomly generated password from a test controller that no longer exists.

**Review by anthonydillon — COMMENTED** [review]
QA complete via a demo.

**Review by anthonydillon — APPROVED** [review]

**Comment by hatched** [comment]
Thanks all!

**Pull request merged**